Jenkinsfile

```diff
@@ -3,45 +3,90 @@
 
 standardProperties()
 
-def runTests(String name, String bedrockCommand, Boolean runAll) {
-  // Clean out the old XML files before running tests, since we junit import *.XML files
-  dir('scratch') {
-    if (isUnix()) {
-      sh "rm -f *.xml"
-    } else {
-      bat "del *.xml"
-    }
-  }
-
-  def command = runAll ? bedrockCommand + ' --ignore-lerna-changed=true' : bedrockCommand
-  def testStatus = exec(script: command, returnStatus: true)
-
-  echo "Writing JUnit results for ${name} on node: $NODE_NAME"
+def runBedrockTest(String name, String command, Boolean runAll, int retry = 0, int timeout = 0) {
+  def bedrockCmd = command + (runAll ? " --ignore-lerna-changed=true" : "")
+  echo "Running Bedrock cmd: ${command}"
+  def testStatus = sh(script: command, returnStatus: true)
   junit allowEmptyResults: true, testResults: 'scratch/TEST-*.xml'
 
-  // If the tests failed (exit code 4) then just mark it as unstable
   if (testStatus == 4) {
     unstable("Tests failed for ${name}")
   } else if (testStatus != 0) {
-    error("Unexpected error running tests for ${name} so passing failure as exit code")
+    if (retry > 0) {
+      echo "Running retry [${retry}] after [${timeout}]"
+      sleep(timeout)
+      runBedrockTest(name, command, runAll, retry - 1, timeout)
+    } else {
+      archiveArtifacts artifacts: 'scratch/TEST-*.xml', onlyIfSuccessful: false
+      error("Unexpected error in ${name} ")
+    }
   }
 }
 
-def runBrowserTests(String name, String browser, String os, Integer bucket, Integer buckets, Boolean runAll) {
+def runHeadlessTests(Boolean runAll) {
+  def bedrockCmd = "yarn grunt headless-auto"
+  runBedrockTest('headless', bedrockCmd, runAll)
+}
+
+def runRemoteTests(String name, String browser, String provider, String platform, String arn, String bucket, String buckets, Boolean runAll, int retry = 0, int timeout = 0) {
+  def awsOpts = " --sishDomain=sish.osu.tiny.work --devicefarmArn=${arn}"
+  def platformName = platform != null ? " --platformName='${platform}'" : ""
   def bedrockCommand =
-    "yarn grunt browser-auto" +
-      " --chunk=400" +
-      " --bedrock-os=" + os +
-      " --bedrock-browser=" + browser +
-      " --bucket=" + bucket +
-      " --buckets=" + buckets;
-
-  runTests(name, bedrockCommand, runAll);
+  "yarn browser-test" +
+    " --chunk=400" +
+    " --bedrock-browser=" + browser +
+    " --remote=" + provider +
+    " --bucket=" + bucket +
+    " --buckets=" + buckets +
+    " --name=" + name +
+    "${provider == 'aws' ? awsOpts : ''}" +
+    "${platformName}"
+    runBedrockTest(name, bedrockCommand, runAll, retry, timeout)
 }
 
-def runHeadlessTests(Boolean runAll) {
-  def bedrockCommand = "yarn grunt headless-auto";
-  runTests("chrome-headless", bedrockCommand, runAll);
+def runTestPod(String cacheName, String name, String browser, String provider, String platform, String bucket, String buckets, Boolean runAll) {
+  return {
+    bedrockRemoteTools.nodeConsumerPod(
```

Comment:
I feel like maybe we should have more review over the waluigi changes that are mostly a black box to editor developers, but that's a discussion for another day